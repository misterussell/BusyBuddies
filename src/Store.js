import User from './models/User';

export default class Store {
  constructor() {
    this.user = new User();
  }
};
