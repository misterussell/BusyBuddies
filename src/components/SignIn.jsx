import React, { useState } from 'react';
import PropTypes from 'prop-types';
import FormValidator from '../models/FormValidator';

function SignIn(props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [signInErr, setSignInErr] = useState({
    isInvalid: false,
    message: '',
  });
  const [emailErr, setEmailErr] = useState({
    isInvalid: false,
    message: '',
  });
  const [passwordErr, setPasswordErr] = useState({
    isInvalid: false,
    message: '',
  });
  const formValidator = new FormValidator([
    {
      field: 'email',
      method: 'isEmpty',
      validWhen: false,
      message: 'Please provide an email address.',
    },
    {
      field: 'email',
      method: 'isEmail',
      validWhen: true,
      message: 'Please enter a valid email address.',
    },
    {
      field: 'password',
      method: 'isEmpty',
      validWhen: false,
      message: 'Please enter a password'
    }
  ]);

  const emailErrDiv = emailErr.isInvalid
    ? (
      <div className="form-field-error">
        <p>{ emailErr.message }</p>
      </div>
    ) : null;

  const passwordErrDiv = passwordErr.isInvalid
    ? (
      <div className="form-field-error">
        <p>{ passwordErr.message }</p>
      </div>
    ) : null;

  const signInErrDiv = signInErr
    ? (
      <div className="sign-in-error">
        { signInErr.message }
      </div>
    )
    : null;

  return (
    <div className="sign-in">
      <form>
        <p>email:</p>
        <input
          value={email}
          onChange={handleEmailChange}
          placeholder="email"
          label="email"
        />
        {
          emailErrDiv
        }
        <p>password:</p>
        <input
          value={password}
          onChange={handlePasswordChange}
          placeholder="password"
          label="password"
        />
        {
          passwordErrDiv
        }
        <button
          onClick={handleSubmit}
          label="submit"
          >sign in</button>
      </form>
      {
        signInErrDiv
      }
    </div>
  );

  function handleEmailChange(e) {
    setEmail(e.target.value);
  };

  function handlePasswordChange(e) {
    setPassword(e.target.value);
  }

  function handleSubmit(e) {
    e.preventDefault();
    const formState = { email, password };
    const validation = formValidator.validate(formState);
    if (validation.isValid) {
      props.user.signIn('max@misterussell.com', 'NewUser1!').then(response => {
        console.log('no error signing in');
        props.user.setAuthenticated();
      }).catch(error => {
        setSignInErr({isInvalid: true, message: error.message});
      });
    } else {
      // - render the validation errors
      // - validatoin.isValid will be false when these errors need to be thrown
      // - form validated errors they stored as keys in the validation object
      // - confirmatoin errors returned from Amazon are stored in the confirmationErr object
      // - first iterate over the form validation errors as these are going to be thrown before the
      // cognito error
      if (validation.email.isInvalid) setEmailErr(validation.email);
      if (validation.password.isInvalid) setPasswordErr(validation.password);
    }
  }
}

SignIn.propTypes = {
  user: PropTypes.object,
};

export default SignIn;
