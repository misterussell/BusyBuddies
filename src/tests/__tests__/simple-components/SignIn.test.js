import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'

import SignIn from '../../../components/SignIn';
import Store from '../../../models/User';

const store = new Store();

beforeEach(() => {
  const div = document.createElement('div');

  ReactDOM.render(<SignIn user={ store.user } />, div);
  ReactDOM.unmountComponentAtNode(div);
});


describe('the SignIn component', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    
    ReactDOM.render(<SignIn user={ store.user } />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders SignIn component correctly', () => {
    const SignInComponent = renderer.create(<SignIn user={ store.user }/>).toJSON();

    expect(SignInComponent).toMatchSnapshot();
  });

});
