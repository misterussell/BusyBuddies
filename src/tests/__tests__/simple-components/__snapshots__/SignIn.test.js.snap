// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the SignIn component renders SignIn component correctly 1`] = `
<div
  className="sign-in"
>
  <form>
    <p>
      email:
    </p>
    <input
      label="email"
      onChange={[Function]}
      placeholder="email"
      value=""
    />
    <p>
      password:
    </p>
    <input
      label="password"
      onChange={[Function]}
      placeholder="password"
      value=""
    />
    <button
      label="submit"
      onClick={[Function]}
    >
      sign in
    </button>
  </form>
  <div
    className="sign-in-error"
  >
    
  </div>
</div>
`;
